### PR TITLE
When the clear_list flag is false, instead of reseting the search lis…

### PR DESCRIPTION
**Pull request by gdimartino**
…t to whatever it was when use_keychain was called, the list is now reset to whatever it currently is, minus the supplied keychain.

### CODE REVIEW

**Comment by richardszalay** [comment]
Relates to #3

File: lib/spare_keys.rb [hunk 1]
```diff
@@ -26,6 +26,15 @@ def self.use_keychain(keychain_path, clear_list = false, type = nil, domain = ni
                 yield if block_given?
             ensure
                 original_keychain = `security #{type}-keychain #{domain_flag} -s #{original_keychain}` if type
+                
+                unless clear_list
+                    # Grab the keychain list as it looks right now in case
+                    # another process has changed it
+                    current_list = `security list-keychains #{domain_flag} | xargs`
+                    # Remove the supplied keychain
+                    original_list = (current_list.split(" ").reject { | item | item.include? keychain_path }).join(" ")
```
Comment (richardszalay) [inline lib/spare_keys.rb]: I have two concerns here:

1. Splitting on space could cause issues if there is a space in the path. I know that OSX tends to push you away from having a space in your username, but can it be done anyway? Or, alternatively, can the default keychain path be changed to somewhere that _does_ have a space? Perhaps transforming the raw (ie. minus xargs) output might be safer
2. `use_keychain` doesn't require a full path to the keychain, so `login` is a perfectly valid value. However, it would also match `my_awesome_login` or  `/alt/path/to/login.keychain`, which would cause unexpected behavior.

**Review by richardszalay — COMMENTED** [review]
Hey Gustavo,

This is great, thanks so much for persisting! I have a few concerns about a number of edge cases that I've added comments to within the file changes in the PR. Specifically they are;

1. The new impl splits the list of keychains by space, but I worry that it's possible for a keychain path to contain a space. Even if it were escaped, it would still break that code.
2. The use of "includes" could provide false positives (ie. "log" matching "login" and thus removing it)

Let me know if you think you can resolve these issues. If not (which I would understand), I'll look at making the changes myself before merging it into the main branch.

Thanks again for taking the time to contribute!

Richard

**Comment by gdimartino** [comment]
Hi Richard,

Great feedback! I can definitely resolve these. Let me give some thought to it and I will try to take care of them sometime this week.

g

"Remember that it doesn't require an awful lot of skill to write a program that a computer understands, the skill is in writing programs that humans understand.” — Robert C. Martin

**Comment by gdimartino** [comment]
Hi Richard, I have committed a change that addresses the first of your concerns by not splitting based on spaces but creating an array from the un-xarged result of the security call. I am struggling with the second one, though. I will need more time. But you can take a look at what I have so far and give feedback on that.

Also, I wasn't sure if I needed to create a new PR or if these commits to my fork will show up as part of the current PR. Please advise if you don't see my commits.

**Comment by gdimartino** [comment]
Hi Richard,

I was wondering if you had had a chance to look at my code that I checked in. I wasn’t able to create a new PR for it, probably because the current PR still exists or something like that… but you should be able to see my checkin none the less.

So I was only able to address the first problem so far, but I think it’s worth looking at. Maybe we can setup some time to discuss the second item, as I am somewhat struggling with the best way to do that.

Let me know what you find!

Gus


Gustavo Di Martino
Software Engineer

Phunware, Inc.
The Software of Things™

(949) 300-5294 — Mobile

www.phunware.com <http://www.phunware.com/>

**Comment by richardszalay** [comment]
Hi Gustavo,

Apologies for the delay in getting back to you. I've been on leave over the holiday period.

What was wrong with the changes you made in 76bd04ac823aac5864a2fad6dd00fada6faa4f62 that made you revert it?

Cheers,
Richard

**Comment by gdimartino** [comment]
Hi Richard,

No problem about the delay, glad you got some time off.

So, the reason I reverted that change is because it caused the unit tests to fail. It’s hard to remember the details now, but the reason that one or more tests were failing (I can’t remember which tests now) is because the use_keychains routine ends up comparing a base file name to a full path containing the base file name in the “reject” handler. So then I tried to work on some way of making the comparison take into account the possibility of having a full file path vs. just the base name and it got kind of complicated. Then when things got super busy here at work, I decided to just turn in what I have so far. I need to set aside some time to finish this fix, but I was hopeful for some insight from you first :-D

Gus

“I liken them to two ropes going through two holes in the ceiling and over a pulley above. If I wish to support myself by them, I must cling to them both. If I cling only to one and not the other, I go down. I read the many teachings of the Bible regarding God's election, predestination, his chosen, and so on. I read also the many teachings regarding 'whosoever will may come' and urging people to exercise their responsibility as human beings. These seeming contradictions cannot be reconciled by the puny human mind. With childlike faith, I cling to both ropes, fully confident that in eternity I will see that both strands of truth are, after all, of one piece.” - Jim Morren, guest lecturer at Calvin Seminary

**Comment by richardszalay** [comment]
Hi Gustavo,

I finally got around to finishing off this PR. I made a few takes, namely having the keychain path expanded so that we can be sure we're removing the correct one. I also updated the unit tests with another edge case.

The changes are all merged now and have been released as 1.2.0

Thanks again for your persistence!

Richard

**Comment by gdimartino** [comment]
That is great news! I will update and give it a whirl.

Thanks so much for helping me out with this, Richard :-)

Gus
